src/hal/lpc17xx/pwm.rs

```diff
@@ -0,0 +1,294 @@
+// Zinc, the bare metal stack for rust.
+// Copyright 2015 Paul Osborne <osbpau@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! PWM Support for the NXP LPC17xx MCUs
+
+use hal::lpc17xx::peripheral_clock::PeripheralClock::PWM1Clock;
+use core::intrinsics::abort;
+
+use self::PWMChannel::*;
+
+
+#[path="../../util/ioreg.rs"]
+#[macro_use] mod ioreg;
+
+const PWM_CLOCK_DIVISOR: u8 = 4;
+
+/// "Channels" correspond to MR1..6 (0 used for period)
+#[allow(missing_docs)]
+#[derive(Clone, Copy)]
+pub enum PWMChannel {
+  CHANNEL0 = 0,  // reserved for period
+  CHANNEL1,
+  CHANNEL2,
+  CHANNEL3,
+  CHANNEL4,
+  CHANNEL5,
+  CHANNEL6,
+}
+
+impl PWMChannel {
+
+  /// Set the match register for this channel to the provided value
+  fn set_mr(&self, value: u32) {
+    // note that for whatever reason, the match registers are fragmented
+    // across two registers.  They aren't even located next to each other
+    // in memory
+    match *self {
+      CHANNEL0 => { reg::PWM1.mr[0].set_value(value); },
+      CHANNEL1 => { reg::PWM1.mr[1].set_value(value); },
+      CHANNEL2 => { reg::PWM1.mr[2].set_value(value); },
+      CHANNEL3 => { reg::PWM1.mr[3].set_value(value); },
+      CHANNEL4 => { reg::PWM1.mr2[0].set_value(value); },
+      CHANNEL5 => { reg::PWM1.mr2[1].set_value(value); },
+      CHANNEL6 => { reg::PWM1.mr2[2].set_value(value); },
+    };
+  }
+}
+
+/// calculate the number of PWM clock ticks for the given number of microseconds
+fn pwm_us_to_ticks(us: u32) -> u32 {
+  let pwm_clock_frequency_hz: u32 = PWM1Clock.frequency();
+  let pwm_clock_frequency_mhz: u32 = pwm_clock_frequency_hz / 1_000_000;
+  return pwm_clock_frequency_mhz * us;
+}
+
+
+#[allow(missing_docs)]
+#[derive(Clone, Copy)]
+pub struct PWM {
```

Comment:
Even with `missing_docs` should still document the actual struct even if the values are implicitly documented